ksql-engine/src/main/java/io/confluent/ksql/function/udf/datetime/StringToTimestamp.java

```diff
@@ -16,35 +16,38 @@
 
 package io.confluent.ksql.function.udf.datetime;
 
-import java.text.DateFormat;
-import java.text.ParseException;
-import java.text.SimpleDateFormat;
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
 
 import io.confluent.ksql.function.KsqlFunctionException;
 import io.confluent.ksql.function.udf.Kudf;
 
 public class StringToTimestamp implements Kudf {
 
-  private DateFormat dateFormat = null;
+  private DateTimeFormatter formatter;
 
   @Override
-  public void init() {
-  }
-
-  @Override
-  public Object evaluate(Object... args) {
+  public Object evaluate(final Object... args) {
     if (args.length != 2) {
       throw new KsqlFunctionException("StringToTimestamp udf should have two input argument:"
                                       + " date value and format.");
     }
+
     try {
-      if (dateFormat == null) {
-        dateFormat = new SimpleDateFormat(args[1].toString());
-      }
-      return dateFormat.parse(args[0].toString()).getTime();
-    } catch (ParseException e) {
-      throw new KsqlFunctionException("Exception running StringToTimestamp(" + args[0] +" , "
+      ensureInitialized(args);
+
+      final LocalDateTime dateTime = LocalDateTime.parse(args[0].toString(), formatter);
```

Comment:
These comments apply to both StringToTimestamp and TimestampToString, though its a bigger issue for StringToTimestamp I think.

LocalDateTime drops information about the time zone that may be present in the string being parsed. I'm not sure what time zone its calculating timestamps for (probably UTC?) but I think the results will be wrong if the time in the string has another zone. Maybe use a ZonedDateTime first and then add in the offset?

DateTimeFormatter reserves all the letters in [A-Za-z]. I've seen some users (including the user who reported the original bug https://github.com/confluentinc/ksql/issues/819) use those letters as delimiters (e.g. 'T' between the date and time). Those queries will likely not work with this implementation.

Reply:
Good spot @rodesai! It looks like if the input date time string has a time zone, this will raise a `java.time.temporal.UnsupportedTemporalTypeException`. Since specifying timezones is perfectly OK, this is a bug. We should add test cases where the input has time zone information.

Reply:
Thanks @rodesai, I'll take a look.

Reply:
Timezones are now handled the same way they were with `SimpleDateFormat`.